### PR TITLE
Allow users to rebind 3pids they own (MSC2229)

### PR DESCRIPTION
**Pull request by anoadragon453**
Implements [MSC2229](https://github.com/matrix-org/matrix-doc/blob/master/proposals/2229-rebind-existing-3pid.md).

If someone owns a 3PID and wants to rebind it, let them.

### CODE REVIEW

File: synapse/rest/client/v2_alpha/account.py [hunk 1]
```diff
@@ -412,12 +412,14 @@ class EmailThreepidRequestTokenRestServlet(RestServlet):
...
         super(EmailThreepidRequestTokenRestServlet, self).__init__()
         self.hs = hs
+        self.auth = hs.get_auth()
         self.config = hs.config
         self.identity_handler = hs.get_handlers().identity_handler
         self.store = self.hs.get_datastore()
 
     @defer.inlineCallbacks
     def on_POST(self, request):
+        requester = yield self.auth.get_user_by_req(request)
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/account.py]: This needs to be optional.

File: synapse/rest/client/v2_alpha/account.py [hunk 2]
```diff
@@ -439,7 +441,8 @@ def on_POST(self, request):
             "email", body["email"]
         )
 
-        if existing_user_id is not None:
+        # Allow this MSISDN to be rebound if the requester owns it
+        if existing_user_id != requester.user.to_string():
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/account.py]: Can we add this check to the bind servlet too for sanity please?

Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: What about `/account/3pid/*/requestToken`?

Comment (erikjohnston) [reply synapse/rest/client/v2_alpha/account.py]: I don't understand the question I'm afraid

Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: Should we add the same check to https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-account-3pid-email-requesttoken and the MSISDN equivalent?

Comment (erikjohnston) [reply synapse/rest/client/v2_alpha/account.py]: I guess? Is that what the MSC says?

Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: yes

Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: So adding the check to the /bind servlet is what's breaking Sytest, in that it didn't used to check for already bound emails but now it does. I suppose this isn't really part of the MSC so should we do it here?

**Comment by anoadragon453** [comment]
fyi we're coming back to this one after all the other privacy PRs land

**Comment by anoadragon453** [comment]
This PR will be superseded by another one that implements [MSC2290](https://github.com/matrix-org/matrix-doc/pull/2290) (if it lands) instead.

**Pull request closed without merging**